cloud_storage: provide in-memory manifest dumps via the Admin API

VladLazar opened the pull request:
This PR introduce a new route to the admin API which dumps the JSON formatted in-memory
manifest for a given partition: `/v1/cloud_storage/manifest/{topic}/{partition}`.

A follow-up PR should extend this route to allow it to dump the uploaded manifest.

Related https://github.com/redpanda-data/redpanda/issues/7459

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.1.x
- [ ] v22.3.x
- [ ] v22.2.x

## Release Notes

### Features
* A new admin API endpoint is added at `/v1/cloud_storage/manifest/{topic}/{partition}` which allows for retrieving the in-memory manifest for a partition in JSON format.

VladLazar commented:
Failures are:
* https://github.com/redpanda-data/redpanda/issues/7418

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/a82622ca642d2d1431a1eea2e54bd0fc51e9b74a..b734055e2b8dc3d756625d2cd42ec8d39d67531d):
* Moved the closing of the stream to make it exception safe

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/b734055e2b8dc3d756625d2cd42ec8d39d67531d..4953caf6bffa58790a4f2c70507173d38dbd4331):
* tweak doc-strings

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/4953caf6bffa58790a4f2c70507173d38dbd4331..61275ee81acfe6fef1140d1b4b7d0c457f5deb92):
* rebased on dev to fix conflicts

andijcr commented on src/v/cloud_storage/partition_manifest.cc:
```diff
@@ -1357,6 +1357,37 @@ void partition_manifest::serialize(std::ostream& out) const {
+          serialized.share(0, serialized.size_bytes()), output);
+        serialized.clear();
+
+        if (iso != _insync_offset) {
+            co_await output.close();
+            throw std::runtime_error(fmt_with_ctx(
+              fmt::format,
+              "Manifest changed duing serialization, in sync offset moved from "
+              "{} to {}",
+              iso,
+              _insync_offset));
+        }
+    }
```
Is this something that will routinely happen?

VladLazar replied:
The insync offset advances with every batch processed by the STM, so if a batch is applied while the manifest is being serialized we'll throw. The alternative here is to lock the manifest while serializing it for the request, but I don't like user input being able to lock up the STM. I suppose I could add some retry logic at the request level, but my idea was for the user to retry this until successful.

VladLazar replied:
Hm. This is actually problematic since the response is chunked, which means that the caller can get a truncated manifest if the insync offset changes.

andijcr replied:
As a user, i would be ok with a truncated response if it does not break the json parser

VladLazar replied:
I ended up taking the manifest lock, but placed an upper bound on the amount of time it can be held (3s).

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/07e742a50c6bbaa0a1ce5bad400ce1aa89f86203..f9c9a631f25f8ca470c8560ac85594e7d31c6fd3):
* acquire manifest lock in bounded way

VladLazar commented:
Failure is:
* https://github.com/redpanda-data/redpanda/issues/9166

VladLazar commented on src/v/redpanda/admin_server.cc:
```diff
@@ -4101,6 +4102,69 @@ admin_server::get_partition_cloud_storage_status(
+                        std::move(part),
+                        [](auto& os, auto& part) mutable {
+                            return part->serialize_manifest_to_output_stream(os)
+                              .finally([part = std::move(part), &os] {
```
I think you're right. I recall struggling a bit with trying to contain the usage of `part` to its home shard, so maybe it's a leftover from that.

VladLazar replied:
Fixed

dotnwat commented on src/v/redpanda/admin_server.cc:
```diff
@@ -4101,6 +4102,69 @@ admin_server::get_partition_cloud_storage_status(
 }
 
+ss::future<std::unique_ptr<ss::httpd::reply>> admin_server::dump_manifest(
+  std::unique_ptr<ss::httpd::request> req,
+  std::unique_ptr<ss::httpd::reply> rep) {
```
this will break in upstream/dev due to changes in seastar: needs to be `ss::http::` scoped instead of `httpd`.

VladLazar replied:
Thanks for the call-out. Will update.

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/f9c9a631f25f8ca470c8560ac85594e7d31c6fd3..1c193be8b9a7224c1d0cb044a37df19638c0d0e4):
* rebased on dev
* Updated types to match new seastar version
* Changed route name to `/v1/cloud_storage/manifest` as discussed with John
* Addressed John's nits.

VladLazar commented:
Changes in [fp-prev](https://github.com/redpanda-data/redpanda/compare/677f7468a860c254af434a65fe69f4978a27a3c3..26978c0c3f72f3c4d0d3fc22d81d6ad76385da02) and [fp-before-prev](https://github.com/redpanda-data/redpanda/compare/1c193be8b9a7224c1d0cb044a37df19638c0d0e4..677f7468a860c254af434a65fe69f4978a27a3c3):
* Fixed merge conflict from botched re-base

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/26978c0c3f72f3c4d0d3fc22d81d6ad76385da02..78d4e5062e78add4ba603b950a5030aa4f352688):
* fix a bug in the cloud storage partition status endpoint that would lead to a null deref when `cloud_storage_enabled=false`.

VladLazar commented:
Failure is:
* https://github.com/redpanda-data/redpanda/issues/9837
